lib/Dialect/SV/Transforms/HWMemSimImpl.cpp

```diff
@@ -260,8 +267,43 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
     // Read Logic
     Value rdata = getMemoryRead(b, reg, addr, addMuxPragmas);
     if (!ignoreReadEnableMem) {
-      Value x = b.create<sv::ConstantXOp>(rdata.getType());
-      rdata = b.create<comb::MuxOp>(en, rdata, x, false);
+      size_t addrRange = 1ULL << addr.getType().getIntOrFloatBitWidth();
+      if (!disableMemRandomization && mem.depth < addrRange) {
+        // If out-of-bounds access is possible, then add a check.
+        auto readWire = b.create<sv::WireOp>(rdata.getType());
+        // Create constant for mem.depth
+        auto maxAddr =
+            b.create<ConstantOp>(b.getIntegerAttr(addr.getType(), mem.depth));
+        // Is enable and in-bounds.
+        auto isEnAndInBounds = b.create<comb::AndOp>(
+            en, b.create<comb::ICmpOp>(comb::ICmpPredicate::ult, addr, maxAddr,
+                                       false));
+        b.create<sv::IfDefOp>(
+            "RANDOMIZE_GARBAGE_ASSIGN",
+            [&]() {
+              auto garbageReg = b.create<sv::VerbatimExprOp>(
+                  rdata.getType(), b.getStringAttr("{{0}}"), ValueRange{},
+                  b.getArrayAttr(hw::InnerRefAttr::get(
+                      op.getNameAttr(),
+                      garbageReadRegister.getInnerSymAttr())));
+              // If enable and is-in-bounds then get memory read, else get
+              // random.
+              b.create<sv::AssignOp>(
+                  readWire, b.create<comb::MuxOp>(isEnAndInBounds, rdata,
+                                                  garbageReg, false));
+            },
+            [&]() {
+              // If enable and is-in-bounds then get memory read, else get X.
```

Comment:
LowerSeqToSV stopped ininializations using verbatim so it would be better to follow in the future (not blocking this PR).

Reply:
I'm not sure if there is an alternative here, the register is getting declared inside the ifder block, and its not live at the use block, that's why the verbatim, I think! WDYT ??
```
  sv.ifdef  "RANDOMIZE_GARBAGE_ASSIGN" {
      %_GARBAGE_READ = sv.reg sym @_GARBAGE_READ  : !hw.inout<i32>
    }
```

Reply:
Yeah, that makes sense. Even though maybe we can do something tricky but not sure it's worth doing:
```mlir
  %garbade_assign_val = sv.wire 
  sv.ifdef  "RANDOMIZE_GARBAGE_ASSIGN" {
      %reg = sv.reg
      sv.assign %garbade_assign_val, %reg
      sv.initial {
         sv.bpassign %reg, RANDOM
      }
    } else {
      sv.assign %garbade_assign_val, sv.constantX
    }
 // use sv.read_inout (%garbade_assign_val) as an out-of-bounds value.
```